Added A2S_PLAYER and A2S_RULES support

w4rum opened the pull request:
- added `A2S_PLAYER` and `A2S_RULES` support (see [Server Queries](https://developer.valvesoftware.com/wiki/Server_queries))

- renamed `A2S_INFO` command from `query_game_server` to `query_server_info`
- added dependency: `deprecation`
- old name of `query_server_info` is still supported but deprecated
- deprecated `return_last_data` as it makes little sense when multiple commands are supported

Work that still needs to be done:
- Update README
- Not quite sure how to define dependencies for PyPI. I've added them to
  `requirements.txt` for now.

w4rum committed:
added A2S_PLAYER support; renamed A2S_INFO command

- added dependency: deprecation
- old name of query_server_info is still supported but deprecated
- deprecated return_last_data as it makes little sense when multiple commands are supported

w4rum committed:
added A2S_RULES support; fixed endianness

w4rum committed:
added string decoding to A2S_RULES command

w4rum committed:
fixed decoding bug in A2S_INFO

- was a regression added by the unified header stripping

w4rum committed:
fixed player name decoding in A2S_PLAYERS

Jason2605 commented:
Firstly, thanks very much for the PR, it's much appreciated, however, when testing the two new methods out locally i seem to be getting an assertion error. It may be something obvious i've missed - i haven't worked on this for a few years now :)

Some context - the server i attempted to connect to was a Rust server i found on a server listing.
```
>>> from steam import SteamQuery
>>> x = SteamQuery("164.132.207.129", 28065)
>>> x.query_server_config()
b'A\x88\xa0\x9f\x08'
b'A\x88\xa0\x9f\x08'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Jason_000/Documents/Stuff/Github/Python/SteamQuery/steam/query.py", line 136, in query_server_config
    assert response.startswith(A2S_RULES_RESPONSE)
AssertionError
>>> x.query_player_info()
b'A\x88\xa0\x9f\x0b'
b'A\xf0\xc7\x86\x0c'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Jason_000/Documents/Stuff/Github/Python/SteamQuery/steam/query.py", line 81, in query_player_info
    assert response.startswith(A2S_PLAYER_RESPONSE)
AssertionError
```

The two binary string outputs are the value of the response before both assertions in both the `query_server_config` and the `query_player_info` methods.

w4rum commented:
I reproduced the error with Rust servers. It seems that the server rejects a few challenge-responses before accepting them. I just don't know why.

```python
>>> b'U\xff\xff\xff\xff'     # Empty challenge response
<<< b'A\xf27W4'              # Challenge
>>> b'U\xf27W4'              # Challenge response
<<< b'A\xf27W4'              # Same challenge again (should be response)

>>> b'U\xff\xff\xff\xff'     # Empty challenge response
<<< b'A\xf27W4'              # Challenge
>>> b'U\xf27W4'              # Challenge response
<<< b'A\xb4LR\x0b'           # Different challenge (should be response)

>>> b'U\xff\xff\xff\xff'     # Empty challenge response
<<< b'A\xb4LR\x0b'           # Challenge
>>> b'U\xb4LR\x0b'           # Challenge response
<<< [Actual rules data]      # Actual data
```
I can not reproduce this behaviour with Squad, ARK or DayZ servers, so this might be a bug in the Rust server implementation. Looking around a bit, other implementations seem to have the same issue:
Excerpt from [python-a2s](https://github.com/Yepoleb/python-a2s/blob/master/a2s/rules.py):
```python
if response_type == A2S_CHALLENGE_RESPONSE:
    if retries >= DEFAULT_RETRIES:
        raise BrokenMessageError(
            "Server keeps sending challenge responses")
    challenge = reader.read_uint32()
    return rules_impl(
        address, timeout, encoding, challenge, retries + 1)
```
I'll add a similar workaround.

Jason2605 commented:
Perfect!

w4rum committed:
added retry mechanism for challenges;

- moved UDP connection handling and error catching into unified decorator method
- added error messages for split packages (unsupported) and servers sending too many retries
- added support for A2S_PLAYERS responses with more than 255 players

w4rum commented:
I added the retry mechanism. I also moved the UDP connection startup/close and error handling into a decorator method to avoid duplicate code.

For some reason, the A2S protocol only uses a signed 8-bit integer for players amounts. Of course this breaks for 300 player Rust servers. I've made the function ignore the player amount in the response and just parse all the available player data.

The player count returned by `query_server_info` will still be wrong but I doubt there's anything we can do about that. The A2S protocol just doesn't support that.

Jason2605 commented:
Hah yeah, i never even noticed that - good spot. I tested the changes against the same rust server and they seemed to work fine. I'm happy to merge this in then - thanks very much for this PR! I'll get round to updating the readme, and creating a new release for this soon enough.
